1467 scholarly works pathway

EricDurante opened the pull request:
Resolves #1467 

@rschenk This isn't finished yet, but I wouldn't mind having your eyeballs on what I'm doing here. For some context, [this part](https://github.com/psu-libraries/scholarsphere/pull/1466) is already done.

One thing to note is that we decided to add some of the validations that formerly only applied to published works to the earlier steps in the deposit process. For example, description and published date are now required to save the "work details"  form. This duplicates the conditional validations in the `WorkVersion` model. I considered removing validations from the model and allowing the forms to handle them entirely, but then I realized that the validations may still need to be implemented at the model level also since there are other ways (apart from the web forms) to create work versions that may also need to be validated (like via the API?). Does that sound right to you?

Otherwise, how does this approach feel to you?

EricDurante committed:
Define the deposit pathway for each type of work

EricDurante committed:
Differentiate work details step in deposit workflow for scholarly works

This also adds some validations to this step for every deposit pathway.

rschenk commented:
> there are other ways (apart from the web forms) to create work versions that may also need to be validated (like via the API?). Does that sound right to you?

Yep, I think this is spot on. I don't know how the API works (that was something developed by Adam and Seth without me) but I know it was used to port all content from SS3 to SS4 so clearly it _is_ creating work-versions outside the context of the deposit form

rschenk commented on app/models/work.rb:
```diff
@@ -212,6 +226,16 @@ def latest_published_version_dois
+      :general
+    end
+  end
+
```
I really like where this is going, the only thing I see is that perhaps the `Work` model should not have any knowledge of the deposit pathway. Perhaps this method should live elsewhere?

EricDurante replied:
Good call. I didn't like this either.

I think I know how I want to refactor it.

EricDurante replied:
https://github.com/psu-libraries/scholarsphere/pull/1480/commits/8fbf1612ade5cc77480517020ddff0762ab9f95a

EricDurante committed:
Fix a couple of code style issues

EricDurante committed:
Refactor logic for determining work deposit pathway

EricDurante committed:
Disable the work type selection field in the review & publish step

Since the appearance and behavior of the forms for depositing a work now
depend on the work type, it would be confusing for a user to be able to
change the type, but not see the change in appearance/behavior until
they submit the form. It seems unlikely that a user would need to change
the work type at this stage anyway, but if they did need to for some
reason, they could return to the first step in the workflow where the
work type is initially set.

EricDurante committed:
Remove some code that I commented but forgot to delete

EricDurante committed:
Differentiate review & publish deposit step for scholarly works

This feels pretty overkill right now. We’ll see if this design bears out
as things get more complicated.

EricDurante commented:
This is ready for a final review, I think.

ajkiessl commented on app/models/work_version.rb:
```diff
@@ -310,6 +314,8 @@ def initial_draft?
            :auto_generated_thumbnail?,
            :thumbnail_url, to: :work
 
+  delegate :deposit_pathway, to: :work, prefix: false
```
Should this be removed?  If I'm understanding correctly, this logic is now held in the WorkDepositPathway class.

EricDurante replied:
Yep - I missed this. Thanks!

ajkiessl commented on spec/models/work_version_spec.rb:
```diff
@@ -246,6 +246,7 @@
   it { is_expected.to delegate_method(:proxy_depositor).to(:work) }
   it { is_expected.to delegate_method(:work_type).to(:work) }
   it { is_expected.to delegate_method(:thumbnail_url).to(:work) }
+  it { is_expected.to delegate_method(:deposit_pathway).to(:work) }
```
Same here.  Can this be removed?

ajkiessl reviewed (changes requested):
Looks great so far, Eric.  I haven't gotten through all the tests yet, but I wanted to drop these change requests in the meantime.

EricDurante committed:
Prune dead code

ajkiessl commented on app/forms/work_deposit_pathway.rb:
```diff
@@ -0,0 +1,151 @@
+                  allow_blank: true,
+                  format: {
+                    with: /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/, # rubocop:disable Layout/LineLength
+                    message: I18n.t('activerecord.errors.work_version.attributes.version_name.format'),
```
The path to this translation is missing `.models`: `I18n.t('activerecord.errors.models.work_version.attributes.version_name.format')`

Also, I'm curious why this validation is here.  It's basically the same as the model level validation except for `allow_blank: true`.

EricDurante replied:
@ajkiessl Thanks for not just reviewing the code but also doing thorough QA - very much appreciated!

You're right. I just needed the form to reflect the underlying model validation here since the rules are identical, but there are better ways to do that than what I did (duplicating the same validation logic).

ajkiessl commented on app/models/work_version.rb:
```diff
@@ -123,7 +123,7 @@ def label(id)
             format: {
               with: /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/, # rubocop:disable Layout/LineLength
-              message: 'Version names must be in semantic version format, ex. 1.0.0',
+              message: I18n.t('format'),
```
This translation didn't work either.  `I18n.t('activerecord.errors.models.work_version.attributes.version_name.format')` worked for me.

EricDurante replied:
Yep - I did this wrong. Good catch!

ajkiessl reviewed (changes requested):
Pretty much everything else looks good.  I just noticed two things below.

EricDurante committed:
Refactor duplication in form validation

Also, fix a broken validation error message translation.

EricDurante committed:
Show all form validation errors at once

…and ensure that any new model-level validations are reflected in each
of the forms.

ajkiessl reviewed (approved):
LGTM thanks!
